File: src/server/handlers/api/internal/sessions/create.rs
use argon2::{
    Argon2, PasswordHash, PasswordVerifier,
    password_hash::rand_core::{OsRng, RngCore},
};
use axum::{Json, extract::State, response::IntoResponse};
use sha2::{Digest, Sha256};

use crate::constants::PASSWORD_MAX_LENGTH;
use crate::server::AppState;

#[derive(serde::Deserialize, utoipa::ToSchema)]
pub struct CreateSessionRequest {
    username: String,
    password: String,
}

#[derive(serde::Serialize, utoipa::ToSchema)]
pub struct CreateSessionResponse {
    access_token: String,
}

pub enum CreateSessionError {
    ServerError,
    InvalidCredentials,
}

impl IntoResponse for CreateSessionError {
    fn into_response(self) -> axum::response::Response {
        match self {
            CreateSessionError::ServerError => (
                axum::http::StatusCode::INTERNAL_SERVER_ERROR,
                "Internal Server Error",
            ),
            CreateSessionError::InvalidCredentials => {
                (axum::http::StatusCode::UNAUTHORIZED, "Invalid credentials")
            }
        }
        .into_response()
    }
}

#[utoipa::path(
    post,
    path = "/api/internal/sessions/create",
    request_body = CreateSessionRequest,
    responses(
        (status = 200, description = "Session created", body = CreateSessionResponse),
        (status = 401, description = "Invalid credentials"),
        (status = 500, description = "Internal server error"),
    )
)]
pub async fn create_session(
    State(state): State<AppState>,
    Json(body): Json<CreateSessionRequest>,
) -> Result<Json<CreateSessionResponse>, CreateSessionError> {
    if body.password.len() > PASSWORD_MAX_LENGTH {
        return Err(CreateSessionError::InvalidCredentials);
    }

    let account = sqlx::query!(
        "SELECT id, hashed_password FROM accounts WHERE name = $1",
        body.username
    )
    .fetch_optional(&state.db)
    .await
    .map_err(|e| {
        tracing::error!(err = %e, "CREATE_SESSION.FAILED_TO_FETCH_ACCOUNT");
        CreateSessionError::ServerError
    })?
    .ok_or(CreateSessionError::InvalidCredentials)?;

    let parsed_hash = PasswordHash::new(&account.hashed_password).map_err(|e| {
        tracing::error!(err = %e, "CREATE_SESSION.FAILED_TO_PARSE_HASH");
        CreateSessionError::ServerError
    })?;

    Argon2::default()
        .verify_password(body.password.as_bytes(), &parsed_hash)
        .map_err(|e| match e {
            argon2::password_hash::Error::Password => {
                tracing::warn!("CREATE_SESSION.INVALID_PASSWORD");
                CreateSessionError::InvalidCredentials
            }
            _ => {
                tracing::error!(err = %e, "CREATE_SESSION.FAILED_TO_VERIFY_PASSWORD");
                CreateSessionError::ServerError
            }
        })?;

    let token_id = uuid::Uuid::new_v4();
    let mut random_bytes = [0u8; 64];
    OsRng.fill_bytes(&mut random_bytes);
    let access_token = format!("{}.{}", token_id, hex::encode(random_bytes));

    let hashed_token = Sha256::digest(access_token.as_bytes()).to_vec();

    sqlx::query!(
        "INSERT INTO access_tokens(id, account_id, persona_id, hashed_token, description, is_super_token) VALUES ($1, $2, NULL, $3, 'Web Session', TRUE)",
        token_id,
        account.id,
        hashed_token,
    )
    .execute(&state.db)
    .await
    .map_err(|e| {
        tracing::error!(err = %e, "CREATE_SESSION.FAILED_TO_INSERT_TOKEN");
        CreateSessionError::ServerError
    })?;

    tracing::info!(account_id = %account.id, "CREATE_SESSION.SUCCESS");

    Ok(Json(CreateSessionResponse { access_token }))
}

File: src/server/handlers/api/internal/sessions/create.ts
import { createHash } from "node:crypto"

import { verify } from "argon2"
import { Hono } from "hono"
import { describeRoute, resolver, validator } from "hono-openapi"
import * as v from "valibot"

import { db } from "../../../../db/index.ts"
import { accessTokensTable } from "../../../../db/schema.ts"

const router = new Hono()

router.post(
    "/",
    describeRoute({
        operationId: "create_session",
        responses: {
            200: {
                description: "ログイン成功",
                content: {
                    "application/json": {
                        schema: resolver(
                            v.object({
                                accessToken: v.string(),
                            })
                        ),
                    },
                },
            },
        },
    }),
    validator(
        "json",
        v.object({
            username: v.string(),
            password: v.pipe(v.string(), v.maxLength(128)),
        })
    ),
    async c => {
        const body = c.req.valid("json")

        const account = await db.query.accountsTable.findFirst({
            columns: {
                id: true,
                hashedPassword: true,
            },
            where: {
                name: body.username,
            },
        })

        if (account == null) {
            return c.text("Invalid username or password", 401)
        }

        const isPasswordValid = await verify(account.hashedPassword, body.password)

        if (!isPasswordValid) {
            return c.text("Invalid username or password", 401)
        }

        const tokenId = crypto.randomUUID()
        const randomBytes = crypto.getRandomValues(new Uint8Array(64))

        const accessToken = `${tokenId}.${randomBytes.toHex()}`
        const hashedToken = createHash("sha256").update(accessToken).digest()

        await db.insert(accessTokensTable).values({
            id: tokenId,
            accountId: account.id,
            description: c.req.header("User-Agent") ?? "Unknown",
            hashedToken,
            isSuperToken: true,
        })

        return c.json({
            accessToken,
        })
    }
)

export default router
